app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,50 @@
+class PasswordResetsController < ApplicationController
+  def new
+    if logged_in?
+      redirect_to root_url 
+      flash[:message] = "You are already logged in." 
+    end 
+  end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email])
+    if @user.present? 
+      token = SecureRandom.urlsafe_base64 
+      @user.update_attributes({ reset_token: token,
+                                reset_digest: digest(token) })
+      UserMailer.password_reset(@user).deliver_now
+      redirect_to root_url 
+      flash[:message] = "Please check your email to reset your password."
+    else 
+      redirect_to new_password_reset_path
+      flash[:danger] = "Please provide a valid email address."
+    end 
+  end 
+
+  def edit
+    @user = User.find_by(email: params[:email])
+    if !(@user && @user.authenticated?(:reset, params[:id]))    
+      flash[:danger] = "Not a valid password reset request."
+      redirect_to new_password_reset_path 
+    end 
+  end
+
+  def update 
+    user = User.find(params[:id])
+    user.update_attributes(user_params)
+    log_in(user)
+    redirect_to user
+  end 
```

Comment:
Sorry for the delay, I'm sort of having trouble making sure those validations work, but I'll figure it out soon enough.